### PR TITLE
ChainReader now subclasses BaseReader

### LINKED ISSUE #3657: Applying Transformation to Chain of Traj doesn't Apply to First Frame of 2nd, 3rd... trajectories
## Expected behavior ##

After creating a universe with multiple trajectory files, and a series of transformations are applied following the protein example. Those transformations are expected to be applied to all frames.

## Actual behavior ##

The first frame of the 2nd, 3rd, etc. trajectories are unwrapped. This is evidenced by writing out the coordinates to the screen, or writing to a pdb file.

It seems to be a discrepancy with the atoms.positions and ts._pos being synced, as the later is used by the Writer.

## Code to reproduce the behavior ##

Run short LJ lammps simulation.
```
units           lj
atom_style      full

region          box block 0 10 0 10 0 10
create_box      2 box bond/types 1 extra/bond/per/atom 1
create_atoms    1 single 5 5 5.5
create_atoms    1 single 5 5 4.5
create_atoms    2 random 200 1234 box

pair_style      lj/cut 4.5
pair_coeff      * * 1.0 1.0
mass            * 1.0

bond_style      harmonic
bond_coeff      1 1000.0 1.0
create_bonds    single/bond 1 1 2

##########
neighbor 2.0 bin
neigh_modify delay 0 every 1 check yes

min_style cg # conjugate gradient (CG) algorithm
minimize 1.0e-4 1.0e-6 1000 10000

###########

variable TK equal 2
variable Pbar equal 0.01
variable seed equal 1234
variable freq equal 1000

velocity all create ${TK} ${seed} rot yes dist gaussian

fix             1 all nvt temp ${TK} ${TK} 0.05

timestep        0.01
reset_timestep 0
thermo          ${freq}

write_data initial.data
dump 1 all custom ${freq} dump_1.lammpstrj id mol type q xu yu zu
run             5000
undump 1

dump 1 all custom ${freq} dump_2.lammpstrj id mol type q xu yu zu
run             5000

```

Import LAMMPS trajectories, center around bonded molecule, and write to pdb.
``` python
from MDAnalysis import transformations as trans
import MDAnalysis as mda

runs = [1,2]
select = "type 1"

data_file = "run0/initial.data"
dumpfile = "run0/dump_{}.lammpstrj"
dump_file = [dumpfile.format(run) for run in runs]

universe = mda.Universe(data_file, dump_file, format="LAMMPSDUMP")
group_target = universe.select_atoms(select)
group_remaining = universe.select_atoms("all") - group_target
box_center = universe.trajectory[0].dimensions[:3] / 2
com = group_target.center_of_mass(unwrap=True)
transforms = [
    trans.unwrap(group_target),
    trans.translate(box_center-com),
    trans.wrap(group_remaining)
]
universe.trajectory.add_transformations(*transforms)

group = universe.select_atoms("all")
with mda.Writer("output_unwrapped.pdb", group.n_atoms) as W:
    for ii,ts in enumerate(universe.trajectory):
        W.write(group)

```

## Current version of MDAnalysis ##

- MDA Version: '2.1.0-dev0'
- Python Version: 3.8.12
- MacOS

### LINKED ISSUE #4008: Some frames (the ones at the trajectory boundaries) are not transformed when using transformations on universes from chainreaders
I am not sure if this is related (i.e. the same underlying cause) as #3343 (there it seems to happen more often than just at the trajectory boundaries, but for the DCD traj below I also have more issues than just at the traj boundaries).
Please let me know if you need any additional info.

## Expected behavior ##

All frames in a trajectory are transformed according to the defined on-the-fly transformations.

## Actual behavior ##

The first frame (frame 0) of all trajectories except the first trajectory one are not transformed (if my transformation depends on an atom group).
Although this seems to be True only for XTC and TRR. The behavior I get If I do the same with the DCD from the testfiles more closely resembles #3343 i.e. every second value is wrong (independent of the trajectory boundaries). See the code example below.

## Code to reproduce the behavior ##

The below code will produce three plots, all of them should be a flat line (we substract the Center of Mass of all atoms in the trafo and plot the center of mass of all atoms). However for XTC and TRR we have one jump at frame 10 (which is the first frame of the second trajectory) and for DCD we have strange oscillations. Adding a third TRR/XTC results in another jump at frame 20. I dont see any correlation for the DCD with the trajectory boundary (at frame 98) except maybe that the oscillattions of the y-coordinate seem even stronger.

``` python
import numpy as np
import matplotlib.pyplot as plt
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import PSF, DCD,  TPR, XTC, TRR

# define our two trafos, the first one works, the second one does not work on all frames
def shift_up(shift=10):
    def wrapped(ts):
        ts.positions[:, 2] += shift
        return ts
    return wrapped

def shift_by_group(ag):
    # this just shifts all atoms in the box by their center of mass,
    # i.e. afterwards the center of mass of all atoms should be [0,0,0]
    def wrapped(ts):
        shift = ag.center_of_mass()
        ts.positions[:] -= shift
        return ts
    return wrapped

# the lists make sure we get the chainreader
universes = [mda.Universe(PSF, [DCD, DCD]), mda.Universe(TPR, [XTC, XTC]), mda.Universe(TPR, [TRR, TRR])]
traj_formats = ["DCD", "XTC", "TRR"]
fig, axs = plt.subplots(3)

for ax, u, name in zip(axs, universes, traj_formats):
    com = np.empty((len(u.trajectory), 3))
    ag = u.atoms
    trafos = (shift_by_group(ag),)
    u.trajectory.add_transformations(*trafos)
    for i, ts in enumerate(u.trajectory):
         com[i] = ag.center_of_mass()
    ax.plot(com, 'o-')
    ax.set_title(name)
ax.set_xlabel("frames")
fig.tight_layout()
```

## Current version of MDAnalysis ##

- Which version are you using? 2.4.2
- Which version of Python (`python -V`)? Python 3.9.7
- Which operating system? linux

EDIT: nicer plot produced.

### PR DESCRIPTION
**Pull request by richardjgowers**
Fixes #4008
Fixes #3657

Changes made in this Pull Request:
 - 

ChainReader currently subclasses `ProtoReader` rather than `ReaderBase`, then attempts to implement many things regarding auxiliaries and transformations itself.  I think it's getting many subtle things wrong with aux and transformations.

[This PR makes `ChainReader` just another subclass of `ReaderBase`, which ultimately simplifies it.](https://github.com/MDAnalysis/mdanalysis/pull/3906/files#r1205817202)

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/coordinates/base.py [hunk 1]
```diff
@@ -664,12 +663,27 @@ class ProtoReader(IOBase, metaclass=_Readermeta):
     #: The appropriate Timestep class, e.g.
     #: :class:`MDAnalysis.coordinates.xdrfile.XTC.Timestep` for XTC.
     _Timestep = Timestep
+    filename: Any
+    convert_units: bool
+    _transformations: list
+    _auxs: dict
 
-    def __init__(self):
+    def __init__(self, filename, convert_units):
```
Comment (hmacdope) [inline package/MDAnalysis/coordinates/base.py]: Does convert units have an accepted default?

File: package/MDAnalysis/coordinates/base.py [hunk 2]
```diff
@@ -781,20 +795,21 @@ def OtherWriter(self, filename, **kwargs):
...
+    @abc.abstractmethod
+    def _read_next_timestep(self, ts=None):
         # Example from DCDReader:
         #     if ts is None:
         #         ts = self.ts
         #     ts.frame = self._read_next_frame(etc)
         #     return ts
-        raise NotImplementedError(
-            "BUG: Override _read_next_timestep() in the trajectory reader!")
+        raise NotImplementedError()
```
Comment (hmacdope) [inline package/MDAnalysis/coordinates/base.py]: Leave the original error message?

File: package/MDAnalysis/coordinates/base.py [hunk 3]
```diff
@@ -1613,7 +1625,11 @@ def copy(self):
 
         return new
 
-    def _read_first_frame(self):  # pragma: no cover
+    def _read_next_timestep(self, ts=None):
+        raise NotImplementedError("SingleFrameReader can't read next frame")
```
Comment (hmacdope) [inline package/MDAnalysis/coordinates/base.py]: Add a test.

File: package/MDAnalysis/coordinates/base.py [hunk 4]
```diff
@@ -1613,7 +1625,11 @@ def copy(self):
...
+
+    @abc.abstractmethod
+    def _read_first_frame(self):
         # Override this in subclasses to create and fill a Timestep
```
Comment (hmacdope) [inline package/MDAnalysis/coordinates/base.py]: I assume  abc.abstractmethod takes care of raising error

File: package/MDAnalysis/coordinates/chain.py [hunk 5]
```diff
@@ -217,9 +217,11 @@ class ChainReader(base.ProtoReader):
 
     """
     format = 'CHAIN'
+    _n_frames: int
```
Comment (hmacdope) [inline package/MDAnalysis/coordinates/chain.py]: I'm missing why this is a class attr rather than attribute?

File: package/MDAnalysis/coordinates/chain.py [hunk 6]
```diff
@@ -450,15 +456,6 @@ def convert_pos_from_native(self, x):
     def convert_pos_to_native(self, x):
         return self.active_reader.convert_pos_to_native(x)
 
-    def copy(self):
```
Comment (hmacdope) [inline package/MDAnalysis/coordinates/chain.py:453]: Why remove copy?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few quick comments. :)

File: package/MDAnalysis/coordinates/chain.py (hunk 5 above)
Comment (richardjgowers) [reply package/MDAnalysis/coordinates/chain.py]: this is just a type annotation

File: package/MDAnalysis/coordinates/chain.py (hunk 6 above)
Comment (richardjgowers) [reply package/MDAnalysis/coordinates/chain.py:453]: it's in the base class, this is removing a faulty version of copy - it won't copy auxiliaries correctly

Comment (hmacdope) [reply package/MDAnalysis/coordinates/chain.py:453]: Roger that.

File: package/MDAnalysis/coordinates/chain.py [hunk 7]
```diff
@@ -621,57 +614,3 @@ def __repr__(self):
...
-            list of all the transformations that will be applied to the coordinates
-
-        See Also
-        --------
-        :mod:`MDAnalysis.transformations`
-        """
-        #Overrides :meth:`~MDAnalysis.coordinates.base.ProtoReader.add_transformations`
-        #to avoid unintended behaviour where the coordinates of each frame are transformed
-        #multiple times when iterating over the trajectory.
-        #In this method, the trajectory is modified all at once and once only.
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/chain.py:655]: This comment suggests that there was a reason for applying transformations to individual trajectories instead of to the ChainReader itself. Did you look into this reason more closely?

File: package/MDAnalysis/coordinates/base.py (hunk 2 above)
Comment (richardjgowers) [reply package/MDAnalysis/coordinates/base.py]: I changed the method to be an `abc.abstractmethod` so it's impossible to reach this code now

**Comment by orbeckst** [comment]
Is it realistic to have this in 2.5 or can we move it to 2.6?

**Comment by IAlibay** [comment]
Moving to a 2.6.0 target

File: package/MDAnalysis/coordinates/chain.py (hunk 7 above)
Comment (richardjgowers) [reply package/MDAnalysis/coordinates/chain.py:655]: test coverage for this sort of behaviour seems good, see: https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/coordinates/test_chainreader.py#L138

File: package/MDAnalysis/coordinates/chain.py [hunk 8]
```diff
@@ -159,7 +158,7 @@ def check_allowed_filetypes(readers, allowed):
                                   "supported for formats: {}".format(allowed))
 
 
-class ChainReader(base.ProtoReader):
+class ChainReader(base.ReaderBase):
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/chain.py:161]: this line change is what's happening in this PR.  Previously by subclassing ProtoReader (which is a more distant ancestor) you had to also reimplement how transformations worked.  

Instead ChainReader now subclasses `ReaderBase` (like a "normal" reader) and now only implements `_read_next_timestep` (like a "normal" Reader).  This means that methods `__next__`, `copy`, `add_transformations` are all done by the base class.

tl;dr less code = less bugs

**Comment by richardjgowers** [comment]
I'm not sure what the pylint failure means, it's not a line I've touched so I'm not going to try and understand it

File: package/MDAnalysis/coordinates/base.py [hunk 9]
```diff
@@ -614,7 +613,7 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         return False  # do not suppress exceptions
 
 
-class _Readermeta(type):
+class _Readermeta(abc.ABCMeta):
```
Comment (IAlibay) [inline package/MDAnalysis/coordinates/base.py:616]: re: pylint failure, do you not need to do a super.__init__ call or something on line 625?

**Review by IAlibay — DISMISSED** [review]
Unfortunately pylint errors are not on diff - we'll need to fix it (or mute it) _before_ merge.

File: package/MDAnalysis/coordinates/base.py (hunk 9 above)
Comment (richardjgowers) [reply package/MDAnalysis/coordinates/base.py:616]: Honestly code like this is very much in the "if it ain't broke don't fix it" category for me

Comment (IAlibay) [reply package/MDAnalysis/coordinates/base.py:616]: 🤷🏽 then mute it I guess?

Comment (richardjgowers) [reply package/MDAnalysis/coordinates/base.py:616]: So remove the linter?

Comment (IAlibay) [reply package/MDAnalysis/coordinates/base.py:616]: no there's a # comment you can use - it's in the pylint docs iirc

**Review by hmacdope — APPROVED** [review]
Great work, thanks @richardjgowers!
